### PR TITLE
Resolving MIME_PART_INVALID_SIZE errors

### PR DESCRIPTION
**Pull request by bcampeau**
Although plenty of examples of this error have been provided, the root cause is a little suspect.

I've taken the following strategy to eliminate possible causes:

1. `strlen` is now only used to calculate the length of the JSON file.
2. `filesize` is used for local attachments.
3. `get_headers` and `Content-Length` are used for remote files.

Although `strlen` is technically binary safe, multi-byte characters could in theory play havoc with it.

Finally, I've added two additional verbose exceptions to be thrown that will highlight the specific file and error that occurred before the API is even called.

### CODE REVIEW

File: tests/apple-push-api/test-class-mime-builder.php [hunk 1]
```diff
@@ -28,5 +28,16 @@ public function testAddJSON() {
 		);
 	}
 
+	public function testInvalidJSON() {
+		$eol      = "\r\n";
+		$name     = 'some-name';
+		$filename = 'article.json';
+		$json     = '';
+		$size     = strlen( $json );
```
Comment (mboynes) [inline tests/apple-push-api/test-class-mime-builder.php]: Looks like `$eol` and `$size` aren't used

File: includes/apple-push-api/class-mime-builder.php [hunk 2]
```diff
@@ -76,7 +78,13 @@ public function add_metadata( $meta ) {
 	 * @access public
 	 */
 	public function add_json_string( $name, $filename, $content ) {
-		return $this->build_attachment( $name, $filename, $content, 'application/json' );
+		return $this->build_attachment(
+			$name,
+			$filename,
+			$content,
+			'application/json',
+			strlen( $content )
```
Comment (mboynes) [inline includes/apple-push-api/class-mime-builder.php]: if you're concerned about multi-byte chars, you could use[ `mb_strlen()`](http://php.net/manual/en/function.mb-strlen.php)

**Comment by mboynes** [comment]
:+1:

File: includes/apple-push-api/class-mime-builder.php (hunk 2 above)
Comment (bcampeau) [inline includes/apple-push-api/class-mime-builder.php]: It breaks it unfortunately.

**Pull request merged**